File: backend/go/post/post_dao.go
package post

import (
	"database/sql"
	"fmt"
)

type PostDao struct {
	Db *sql.DB
}

func (d *PostDao) CreatePost(post *Post) error {
	sql := "INSERT INTO post (id, date, title, content) VALUES (?, ?, ?, ?)"
	_, err := d.Db.Exec(sql, post.ID, post.Date, post.Title, post.Content)
	return err
}

func (d *PostDao) GetPost(id *string) (Post, error) {
	sql := "SELECT * FROM post WHERE id = ? LIMIT 1"
	rows, err := d.Db.Query(sql, id)

	var noop Post
	if err != nil {
		return noop, err
	}

	defer rows.Close()
	var post Post

	nextRow := rows.Next()

	if !nextRow {
		return noop, fmt.Errorf("Post not found")
	}

	if err := rows.Scan(&post.Date, &post.Content, &post.ID, &post.Title); err != nil {
		return noop, err
	}

	if err := rows.Err(); err != nil {
		return noop, err
	}

	return post, nil
}

func (d *PostDao) GetAllPosts() ([]Post, error) {
	sql := "SELECT * FROM post ORDER BY date DESC"
	rows, err := d.Db.Query(sql)

	var noop []Post
	if err != nil {
		return noop, err
	}

	var posts []Post

	defer rows.Close()
	for rows.Next() {
		var post Post

		if err := rows.Scan(&post.Date, &post.Content, &post.ID, &post.Title); err != nil {
			return noop, err
		}

		posts = append(posts, post)
	}

	if err := rows.Err(); err != nil {
		return noop, err
	}

	return posts, nil
}

func (d *PostDao) UpdatePost(post *Post) error {
	sql := "UPDATE post SET title = ?, content = ? WHERE id = ?"
	_, err := d.Db.Exec(sql, post.Title, post.Content, post.ID)
	return err
}

func (d *PostDao) DeletePost(id *string) error {
	sql := "DELETE FROM post WHERE id = ?"
	_, err := d.Db.Exec(sql, id)
	return err
}

File: backend/java/src/main/java/com/michaelyi/personalwebsite/post/PostDao.java
package com.michaelyi.personalwebsite.post;

import java.util.List;
import java.util.Optional;

import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Repository;

@Repository
public class PostDao {
    private final JdbcTemplate template;
    private final PostRowMapper mapper;

    public PostDao(JdbcTemplate template, PostRowMapper mapper) {
        this.template = template;
        this.mapper = mapper;
    }

    public void createPost(Post post) {
        String sql = """
                INSERT INTO post (id, date, title, content)
                VALUES (?, ?, ?, ?)
                """;
        template.update(
                sql,
                post.getId(),
                post.getDate(),
                post.getTitle(),
                post.getContent());
    }

    public Optional<Post> getPost(String id) {
        String sql = "SELECT * FROM post WHERE id = ? LIMIT 1";
        return template.query(sql, mapper, id).stream().findFirst();
    }

    public List<Post> getAllPosts() {
        String sql = "SELECT * FROM post ORDER BY date DESC";
        return template.query(sql, mapper);
    }

    public void updatePost(Post post) {
        String sql = """
                UPDATE post SET title = ?, content = ? WHERE id = ?
                """;
        template.update(sql, post.getTitle(), post.getContent(), post.getId());
    }

    public void deletePost(String id) {
        String sql = "DELETE FROM post WHERE id = ?";
        template.update(sql, id);
    }

    public void deleteAllPosts() {
        String sql = "DELETE FROM post";
        template.update(sql);
    }
}
